### ISSUE TITLE
Implement thumbnail creation service

### ISSUE THREAD

**Issue by giohappy** [geosolutions-it/geonode#648]
The current creation of thumbnails for layers and maps is quite a mess, and the following points are especially problematic:

- there are multiple implementations of thumbnail rendering of layers. At least the following two are implemented:
   - [create_gs_thumbnail_geonode](https://github.com/GeoNode/geonode/blob/d712fe93bc1bacda3cb22e521f531bb5d0bbabec/geonode/layers/utils.py#L1094) which apparently is only used by geoserver async signals signals 
   - [layer_thumbnail](https://github.com/GeoNode/geonode/blob/1a94bbe43e553a4b3155260c925a1f4f9919731c/geonode/layers/views.py#L1472)
- both [layer_thumbnail](https://github.com/GeoNode/geonode/blob/1a94bbe43e553a4b3155260c925a1f4f9919731c/geonode/layers/views.py#L1472) and [map_thumbnail](https://github.com/GeoNode/geonode/blob/8448190b6a105f4b9fe94276e442f8a877606de3/geonode/maps/views.py#L1310) [use the pritng Geoserver module](https://github.com/GeoNode/geonode/blob/d909a1d1e50a765f1b849b9657377e07db83cf1d/geonode/geoserver/helpers.py#L2068) for the actaul rendering. This module is stale and quite a black box. Probably [this code](https://github.com/planetfederal/geoserver-exts/tree/master/printng) reflects the module that we're [bundling inside the Geoserver build](https://github.com/GeoNode/geoserver-geonode-ext/blob/2.17.x/geoserver/pom.xml#L427) for GeoNode
- the creation of the thumbnail for a map apparently takes different paths in case the thumbnail is generated from the Django template pages or from the "Save" action inside the GeoNode MapStore client. Different thumbnails (different zoom level and positioning) are generated for the two cases

We want to move away from the pritng module and implement the rendering inside a cleaner backend Python module, with a clear and unqiue interface that will be adopted by any client method. This module could become the basis for the replacement of the printing module too.

This module will be in charge of:

- receiving a configuration for the map to be rendered (list of layers, bbox, dimensions, etc.)
- generating the requests (WM(T)S, TMS) to obtain the images for each layer
- composing the final rendered image

Notice that part of the heavy work (calculation of tiles, URL for tile requests, etc.) is already [done on the Python side](https://github.com/GeoNode/geonode/blob/d909a1d1e50a765f1b849b9657377e07db83cf1d/geonode/geoserver/helpers.py#L2197). 

Several helper libraries already exist for Python for composiion of images, ranging from general purpose like [Pillow](https://pillow.readthedocs.io/en/stable/) to domain centric as [Mapnik](https://mapnik.org/).
The backend lib will be selected after having clarified the detaild of the requirements.

**Comment by giohappy** [comment]
@simboss please add any comment to the desciprion above. More details will be added after the invetigation from @bieganowski will be completed.

**Comment by giohappy** [comment]
@ElenaGallo the mechanism underlying the creation of the thumbnails have been rewritten from scratch. We need to test it thoroughly, all over the places and situations where thumbnails are being created / updated:

 - creation of new resources, in particular layers and maps
 - update of thumbnail from the metadata wizard form
 - update of thumbnail through Edit -> (thumbnail) Set
 - update of thumbnail through Edit -> (thumbnail) Upload

In case expand the tests spreadsheets.

@afabiani should we also test it for the importlayers and updatelayers commands?

**Comment by ElenaGallo** [comment]
@giohappy I updated the [spreadsheet](https://docs.google.com/spreadsheets/d/14Ydroab6rWThGLe9kg3vpEXofQyzudF2vLdTS_ZH7jg/edit?usp=sharing) with the new tests done on the thumbs and the following tests fail:

**1_** Opening the metadata wizard, the thumbnail options are not displayed correctly for a moment
![thumb1](https://user-images.githubusercontent.com/56537133/111604055-36aca500-87d5-11eb-9163-0d06e24bffc1.gif)

*How to Reproduce*
- Login as admin
- Click on Data > Layers 
- Select a layer > Click on Editing Tools
- Click on Wizard under column Metadata

*Current Result*
The thumbnail options are not displayed correctly for a moment

**2_** By saving the map, for the first time, it should be possible to upload a thumb
 (as it already happens with stories)
![thumb2](https://user-images.githubusercontent.com/56537133/111604066-3a402c00-87d5-11eb-81ae-0fc389c70410.gif)

*How to Reproduce*
- Login as admin
- Click on Maps > Create Map
- From Options menu select Catalog > Add a layer
- From Options menu select Save as
- Insert a thumbnails and a Name 
- Click on Save as
- Go back to Maps list 

*Current Result*
The new saved map does not have the thumb loaded

**Comment by giohappy** [comment]
@ElenaGallo the first glitch is not related to this work. 
@bieganowski please check the second one.

**Comment by bieganowski** [comment]
ok, the issue is there and in fact is quite simple - the logic inherited from the original implementation expects a thumbnail name to have `.png` extension. Loading any other file format, even with `overwrite=False` (as in the case of Maps creation), causes regeneration of the thumbnail, since the generation script doesn't recognize the thumbnail.

**Comment by giohappy** [comment]
@bieganowski what about the upload done from the resource edit window's "Upload" button? Does it follow a different path?

![image](https://user-images.githubusercontent.com/571129/111789132-1f4ae600-88c1-11eb-9126-847d68e98c2c.png)

**Comment by bieganowski** [comment]
@giohappy indeed, this one does not use `mapstore` at all, it targets `upload_thumbnail` endpoint from `geonode.base` app, so it does not trigger thumbnail creation at all

**Comment by bieganowski** [comment]
issue created in the upstrem: https://github.com/GeoNode/geonode/issues/7114
